# 03_EDA_ANALISIS_DESCRIPTIVO.ipynb
# Objetivo: análisis descriptivo inicial del dataset ya limpio y transformado

# Importación de librerias y configuración

In [24]:
import os
from pathlib import Path
import warnings
warnings.filterwarnings("ignore")

import numpy as np
import pandas as pd

# Solo Matplotlib (sin seaborn en esta fase de verificación)
import matplotlib.pyplot as plt

# Reproducibilidad
RANDOM_SEED = 42
np.random.seed(RANDOM_SEED)

print("Librerías cargadas correctamente.")

# --- Opciones de pandas para ver mejor las tablas ---
pd.set_option("display.max_rows", 50)
pd.set_option("display.max_columns", 200)
pd.set_option("display.width", 140)
pd.set_option("display.float_format", lambda x: f"{x:,.2f}")

# --- Opciones de pandas para ver mejor las tablas ---
pd.set_option("display.max_rows", 50)
pd.set_option("display.max_columns", 200)
pd.set_option("display.width", 140)
pd.set_option("display.float_format", lambda x: f"{x:,.2f}")

Librerías cargadas correctamente.


## Cargo dataset limpio y transformado

In [25]:

# Ruta del archivo CSV limpio y transformado
ruta_dataset = "../DATA/DATA_OUTPUT/dataset_limpio_y_transformado.csv"

# Cargar el dataset
df = pd.read_csv(ruta_dataset, encoding="utf-8")

# Mostrar dimensiones
print(f"✅ Dataset cargado correctamente.")
print(f"Filas: {df.shape[0]:,}  |  Columnas: {df.shape[1]}")

# Vista rápida de las primeras filas
df.head()


✅ Dataset cargado correctamente.
Filas: 10,000  |  Columnas: 34


transaction_id                           customer_id                   name                   email  \
0  d97ea0b8-1869-4a78-a0cd-6356e6bf5feb  f21d2833-c18e-4d66-9943-97c46f5591ad    Christopher Gilmore       qmercer@gmail.com   
1  6c330754-0fdb-4f31-8cf7-7513d4e39f3a  a3b3e11b-7775-426c-a949-7db69b9381ec        William Barrett    elaine44@hotmail.com   
2  6238ad32-3b74-443c-8278-ace4a2f155a9  9acfe8a8-0a93-4d58-bdc8-c078ef635de6         Linda Williams   woodkaren@hotmail.com   
3  9e0d4019-f135-4738-8314-b1a046fb899f  45e1ded6-e770-4f42-b196-9c966a6e9e59       Vincent Saunders     nancy83@hotmail.com   
4  ec9c235c-b9d1-4ad2-8e5f-7252ffa748e7  2d4cfaec-1127-429e-bdbb-9faa2e33c08b  Mrs. Lindsey Allen Md  browndavid@hotmail.com   

       phone_clean      dominio  risk_score risk_level  created_at  customer_days_active transaction_date transaction_hour  hour  year  \
0     866884634513    gmail.com        0.69      Medio  2025-09-05                   242       2025-01-06         19:06:08     0  2025   
1  100033715734463  hotmail.com        0.27       Bajo  2024-11-03                   141       2025-03-24         23:30:54     0  2024   
2  936376652159352  hotmail.com        0.16       Bajo  2025-03-23                   158       2024-10-16         07:26:32     0  2025   
3       2779202543  hotmail.com        0.58      Medio  2025-05-12                   132       2024-12-31         08:37:49     0  2025   
4   17225903259599  hotmail.com        0.55      Medio  2024-10-20                   292       2025-08-08         17:59:21     0  2024   

   month  day weekday month_year region_normalized                                 country_normalized  is_international  amount currency  \
0      9    5  Friday    2025-09        California                                               Togo                 1  124.96      EUR   
1     11    3  Sunday    2024-11             Texas                                       Saint Martin                 1  151.69      EUR   
2      3   23  Sunday    2025-03              Utah                                             Jersey                 1  160.14      EUR   
3      5   12  Monday    2025-05        New Mexico  British Indian Ocean Territory (Chagos Archipe...                 1   77.03      EUR   
4     10   20  Sunday    2024-10            Kansas                              Sao Tome and Principe                 1  139.38      EUR   

                      merchant merchant_category   entry_mode transaction_result                       arn   card_type  card_bin6  \
0                Brown-Gilbert        restaurant       wallet           approved  37a7048c3a5045f9b8baf73c  MASTERCARD     423952   
1    Davis, Erickson and Davis       electronics    magstripe           approved  f512e83aa9eb48d98b66c6ac        AMEX     431620   
2  Anderson, Barnett and Lloyd            gaming    ecommerce           declined  a6571d209527484da6c0759e  MASTERCARD     477500   
3       Wise, Larson and Brown        restaurant  contactless           declined  ccc1771d83c848038a953797        VISA     468119   
4               Maldonado-Beck       electronics         chip           approved  b0a54d13ca094472aa652ae7        VISA     469023   

   card_last4          card_masked card_expiry  is_fraud  
0        6802  XXXX-XXXX-XXXX-6802       03/24         0  
1        3027  XXXX-XXXX-XXXX-3027       10/26         0  
2        7390  XXXX-XXXX-XXXX-7390       02/25         0  
3        6770  XXXX-XXXX-XXXX-6770       01/29         0  
4        9432  XXXX-XXXX-XXXX-9432       02/29         0

### Creo carpeta de salida para el análisis del EDA

In [26]:
# Ruta base (sube un nivel desde la carpeta NOTEBOOKS)
base_dir = Path("..") / "DATA" / "DATA_OUTPUT"

# Crear carpeta para guardar los resultados del análisis
carpeta_eda = base_dir / "EDA"
carpeta_eda.mkdir(parents=True, exist_ok=True)

print(f"Carpeta creada o ya existente: {carpeta_eda.resolve()}")


Carpeta creada o ya existente: C:\Users\Propietario\Desktop\PROYECTO_FINAL\Proyecto_final_Data_Analytics\DATA\DATA_OUTPUT\EDA


## Chequeo básico de estrusctura y nulos

In [27]:
# Chequeo básico del dataset: tamaño, columnas, tipos y nulos

print("Dimensiones -> Filas, Columnas:", df.shape)
print("\nColumnas:\n", df.columns.tolist(), "\n")

print("=== df.info() ===")
df.info()

# Resumen de nulos (solo las columnas que tengan)
nulos = df.isna().sum().sort_values(ascending=False)
nulos = nulos[nulos > 0]

if not nulos.empty:
    resumen_nulos = (
        nulos.to_frame("nulos")
             .assign(pct=lambda t: (t["nulos"] / len(df) * 100).round(2))
    )
    display(resumen_nulos)
    # Guardar en la carpeta de salida del EDA
    (carpeta_eda / "resumen_nulos.csv").write_text(
        resumen_nulos.to_csv(index=True, encoding="utf-8-sig")
        or ""  # write_text requiere una string; usamos el retorno de to_csv
    )
    print("Resumen de nulos guardado en:", (carpeta_eda / "resumen_nulos.csv").resolve())
else:
    print("No hay columnas con nulos.")


Dimensiones -> Filas, Columnas: (10000, 34)

Columnas:
 ['transaction_id', 'customer_id', 'name', 'email', 'phone_clean', 'dominio', 'risk_score', 'risk_level', 'created_at', 'customer_days_active', 'transaction_date', 'transaction_hour', 'hour', 'year', 'month', 'day', 'weekday', 'month_year', 'region_normalized', 'country_normalized', 'is_international', 'amount', 'currency', 'merchant', 'merchant_category', 'entry_mode', 'transaction_result', 'arn', 'card_type', 'card_bin6', 'card_last4', 'card_masked', 'card_expiry', 'is_fraud'] 

=== df.info() ===
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 34 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   transaction_id        10000 non-null  object 
 1   customer_id           10000 non-null  object 
 2   name                  10000 non-null  object 
 3   email                 10000 non-null  object 
 4   phone_clean           

## Compruebo registros duplicados

In [28]:
# Paso 4️⃣ - Comprobación de duplicados
duplicados = df.duplicated().sum()
print(f"Número de registros duplicados en el dataset: {duplicados}")

# Si hay duplicados, los mostramos (solo los primeros 5)
if duplicados > 0:
    display(df[df.duplicated()].head())
else:
    print("No hay registros duplicados.")


Número de registros duplicados en el dataset: 0
No hay registros duplicados.


## Estadísticas descriptivas

In [29]:
# Resumen numérico general
resumen_numerico = df.describe().T  # Transpuesto para que se vea mejor
display(resumen_numerico)

# Guardar el resumen en la carpeta EDA
(resumen_numerico
    .to_csv(carpeta_eda / "resumen_estadistico.csv", encoding="utf-8-sig"))

print("Resumen estadístico guardado en:", (carpeta_eda / "resumen_estadistico.csv").resolve())


count                   mean                    std           min              25%                  50%  \
phone_clean          10,000.00 203,116,852,426,307.09 436,725,639,314,863.62 49,918,491.00 7,982,948,304.75 8,538,969,503,874.00   
risk_score           10,000.00                   0.51                   0.29          0.00             0.27                 0.51   
customer_days_active 10,000.00                 120.38                  85.24          0.00            48.00               106.00   
hour                 10,000.00                   0.00                   0.00          0.00             0.00                 0.00   
year                 10,000.00               2,024.79                   0.41      2,024.00         2,025.00             2,025.00   
month                10,000.00                   6.43                   3.39          1.00             4.00                 6.00   
day                  10,000.00                  15.49                   8.95          1.00             7.75                16.00   
is_international     10,000.00                   1.00                   0.03          0.00             1.00                 1.00   
amount               10,000.00                 189.61                  94.59          2.22           112.40               190.30   
card_bin6            10,000.00             475,021.31              43,458.45    400,024.00       437,798.00           474,909.50   
card_last4           10,000.00               5,046.61               2,890.10          0.00         2,554.75             5,086.00   
is_fraud             10,000.00                   0.01                   0.10          0.00             0.00                 0.00   

                                        75%                      max  
phone_clean          128,072,547,652,563.00 1,993,280,326,788,724.00  
risk_score                             0.75                     1.00  
customer_days_active                 180.00                   363.00  
hour                                   0.00                     0.00  
year                               2,025.00                 2,025.00  
month                                  9.00                    12.00  
day                                   23.00                    31.00  
is_international                       1.00                     1.00  
amount                               263.04                   579.41  
card_bin6                        512,702.75               549,997.00  
card_last4                         7,563.25                 9,999.00  
is_fraud                               0.00                     1.00

Resumen estadístico guardado en: C:\Users\Propietario\Desktop\PROYECTO_FINAL\Proyecto_final_Data_Analytics\DATA\DATA_OUTPUT\EDA\resumen_estadistico.csv


### 📊 Análisis general del resumen descriptivo

#### ✅ Variables normales

- **`risk_score`** → Promedio ≈ **0.51**, desviación **0.29**, entre 0 y 1 → ✅ *normal y coherente*.  
- **`customer_days_active`** → Media ≈ **120 días**, con máx. **363** → *correcto (equivalente a un año aprox.)*.  
- **`amount`** → Media ≈ **189 €**, máx. **579 €**, mínimos > 2 € → *valores razonables*.  
- **`is_international`** → Media **1**, indica que la mayoría de las transacciones son *internacionales* (ya se observó en pasos previos).  
- **`is_fraud`** → Media **0.01**, aproximadamente **1 % de fraude**, ideal para un dataset con proporción realista.  

---

#### ⚠️ Variables con valores que no tienen sentido numérico

Estas variables deben **excluirse de los análisis estadísticos** porque no representan magnitudes medibles:

- **`phone_clean`**, **`card_bin6`**, **`card_last4`** → *son identificadores o códigos, no variables cuantitativas.*  
  > Ejemplo: el valor medio o la desviación estándar de un número de teléfono o tarjeta carece de significado estadístico.  
  ✅ Estas columnas se excluirán de los cálculos posteriores.

- **`hour`** → Aparece todo en **0**, probablemente porque es una columna de tipo *hora* (`time`) que `describe()` no interpreta correctamente.  
  Más adelante se realizará un **resumen específico de variables temporales** (`hour`, `month`, `day`, etc.).

---

✅ **Conclusión:**  
El dataset presenta valores coherentes y bien distribuidos, lo que permite avanzar al siguiente bloque del análisis descriptivo.


## Análisis descriptivo agrupado por categorías clave

In [30]:
# Número de casos por clase
print("Distribución de fraude:")
print(df["is_fraud"].value_counts(normalize=True).mul(100).round(2).astype(str) + " %")

# Media de las principales variables numéricas por tipo de fraude
comparativa_fraude = (
    df.groupby("is_fraud")[["amount", "risk_score", "customer_days_active"]]
      .mean()
      .round(2)
)

display(comparativa_fraude)

# Guardar resultado en carpeta EDA
comparativa_fraude.to_csv(carpeta_eda / "comparativa_fraude.csv", encoding="utf-8-sig")

print("Comparativa guardada en:", (carpeta_eda / "comparativa_fraude.csv").resolve())


Distribución de fraude:
is_fraud
0    98.92 %
1     1.08 %
Name: proportion, dtype: object


amount  risk_score  customer_days_active
is_fraud                                          
0         189.67        0.51                120.45
1         184.33        0.48                114.01

Comparativa guardada en: C:\Users\Propietario\Desktop\PROYECTO_FINAL\Proyecto_final_Data_Analytics\DATA\DATA_OUTPUT\EDA\comparativa_fraude.csv


### 📊 Interpretación de resultados — Comparativa de fraude

#### 📈 Distribución general
- **Transacciones no fraudulentas (`is_fraud = 0`)**: 98.92 %  
- **Transacciones fraudulentas (`is_fraud = 1`)**: 1.08 %  

Esta proporción confirma un **desbalance realista** en los datos, donde el fraude representa una fracción muy pequeña del total de operaciones.  
Este comportamiento es habitual en contextos financieros y servirá para probar la eficacia de los análisis y modelos posteriores.

---

#### 💰 Comparativa de valores medios

| Variable | Sin fraude | Con fraude | Diferencia destacable |
|-----------|-------------|-------------|-------------------------|
| **amount** | 189.67 € | 184.33 € | Las operaciones fraudulentas muestran un **importe medio ligeramente inferior**, por lo que el fraude no se concentra necesariamente en importes altos. |
| **risk_score** | 0.51 | 0.48 | El nivel de riesgo es **un poco menor en fraudes**, lo que podría indicar que el sistema de puntuación no siempre detecta correctamente los casos sospechosos. |
| **customer_days_active** | 120.45 | 114.01 | Los clientes involucrados en fraude tienden a ser **ligeramente más recientes**, lo que sugiere que las cuentas nuevas podrían tener más probabilidad de actividad fraudulenta. |

---

#### 🧭 Conclusión
En conjunto, los resultados indican que el fraude **no se caracteriza necesariamente por importes más altos ni por puntuaciones de riesgo más elevadas**.  
Sin embargo, la menor antigüedad de los clientes en los casos de fraude podría ser un patrón relevante para explorar en fases posteriores del análisis.

Este tipo de información será útil para:
- Ajustar las reglas de detección de riesgo.  
- Analizar más adelante por **categoría de comercio**, **modo de entrada** o **país** para detectar patrones más claros.  

---

📁 Archivo generado:  
`DATA/DATA_OUTPUT/EDA/comparativa_fraude.csv`
